### PR TITLE
Handle negative free disk space in deciders

### LINKED ISSUE #48380: IllegalArgumentException: Values less than -1 bytes are not supported on DiskThresholdDecider
**Elasticsearch version** (`bin/elasticsearch --version`): 7.4.0

**Plugins installed**: [repository-s3]

**JVM version** (`java -version`): 
```
[root@c1be5a9fd961 /]# /elasticsearch/jdk/bin/java --version
openjdk 13 2019-09-17
OpenJDK Runtime Environment AdoptOpenJDK (build 13+33)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 13+33, mixed mode, sharing)
```

**OS version** (`uname -a` if on a Unix-like system):
```
Linux c1be5a9fd961 4.15.0-1027-aws #27-Ubuntu SMP Fri Nov 2 15:14:20 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```
running on ESS

**Description of the problem including expected versus actual behavior**:
Suddenly during a plan migration, plans started to fail due to the exception:
```
[instance-0000000104] unexpected failure during [cluster_reroute(async_shard_fetch)], current state version [1212336]
java.lang.IllegalArgumentException: Values less than -1 bytes are not supported: -192978829196b
    at org.elasticsearch.common.unit.ByteSizeValue.<init>(ByteSizeValue.java:72) ~[elasticsearch-7.4.0.jar:7.4.0]
    at org.elasticsearch.common.unit.ByteSizeValue.<init>(ByteSizeValue.java:67) ~[elasticsearch-7.4.0.jar:7.4.0]
    at org.elasticsearch.cluster.routing.allocation.decider.DiskThresholdDecider.canRemain(DiskThresholdDecider.java:312) ~[elasticsearch-7.4.0.jar:7.4.0]
    at org.elasticsearch.cluster.routing.allocation.decider.AllocationDeciders.canRemain(AllocationDeciders.java:108) ~[elasticsearch-7.4.0.jar:7.4.0]
    at org.elasticsearch.cluster.routing.allocation.allocator.BalancedShardsAllocator$Balancer.decideMove(BalancedShardsAllocator.java:668) ~[elasticsearch-7.4.0.jar:7.4.0]
    at org.elasticsearch.cluster.routing.allocation.allocator.BalancedShardsAllocator$Balancer.moveShards(BalancedShardsAllocator.java:628) ~[elasticsearch-7.4.0.jar:7.4.0]
    at org.elasticsearch.cluster.routing.allocation.allocator.BalancedShardsAllocator.allocate(BalancedShardsAllocator.java:123) ~[elasticsearch-7.4.0.jar:7.4.0]
    at org.elasticsearch.cluster.routing.allocation.AllocationService.reroute(AllocationService.java:405) ~[elasticsearch-7.4.0.jar:7.4.0]
    at org.elasticsearch.cluster.routing.allocation.AllocationService.reroute(AllocationService.java:370) ~[elasticsearch-7.4.0.jar:7.4.0]
    at org.elasticsearch.cluster.routing.BatchedRerouteService$1.execute(BatchedRerouteService.java:112) ~[elasticsearch-7.4.0.jar:7.4.0]
    at org.elasticsearch.cluster.ClusterStateUpdateTask.execute(ClusterStateUpdateTask.java:47) ~[elasticsearch-7.4.0.jar:7.4.0]
    at org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:702) ~[elasticsearch-7.4.0.jar:7.4.0]
    at org.elasticsearch.cluster.service.MasterService.calculateTaskOutputs(MasterService.java:324) ~[elasticsearch-7.4.0.jar:7.4.0]
    at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:219) [elasticsearch-7.4.0.jar:7.4.0]
    at org.elasticsearch.cluster.service.MasterService.access$000(MasterService.java:73) [elasticsearch-7.4.0.jar:7.4.0]
```

`instance-0000000104` is the master node

**Steps to reproduce**:

Sorry, it's the first and only time I have seen this, so I have no steps.

**Provide logs (if relevant)**:
The exception are up there.

Additionally, after @ywelsch suggestion, I enabled the DEBUG level on `org.elasticsearch.cluster.routing.allocation.decider` which revealed:

```
[instance-0000000106] less than the required 0b free bytes threshold (-194290383982 bytes free) on node aIG71toLQTeY-FIvHziBag, shard cannot remain
```

I identified that node using the _cluster/state and restarted it and the problem seems to be gone.

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today it is possible that the total size of all relocating shards exceeds the
total amount of free disk space. For instance, this may be caused by another
user of the same disk increasing their disk usage, or may be due to how
Elasticsearch double-counts relocations that are nearly complete particularly
if there are many concurrent relocations in progress.

The `DiskThresholdDecider` treats negative free space similarly to zero free
space, but it then fails when rendering the messages that explain its decision.
This commit fixes its handling of negative free space.

Fixes #48380

### CODE REVIEW

**Review by ywelsch — APPROVED** [review]
I'm not sure whether we would like `DiskUsage.freeBytes()` to ever return a negative value. Its serialization as a `VLong` e.g. suggests that that value can never be negative. I know that we only expect this to be negative if we have locally create a `DiskUsage` instance that is not serialized, but that seems trappy either way. I wonder if we should instead add a new class that represents a DiskUsage object with relocations taken into account and use that object for the decision making and logging here. We can do this in a follow-up though.

**Comment by DaveCTurner** [comment]
`DiskUsageTests` mention cases where the disk usage is negative, although looking at `ESFileStore` I don't think that this can really happen any more. But you're certainly right that negative disk usage stats can't be serialised and this seems bad.

I added a new class in 9ea918f.

**Comment by ywelsch** [comment]
> DiskUsageTests mention cases where the disk usage is negative, although looking at ESFileStore I don't think that this can really happen any more.

Should we fix the tests then?

I would like to have these assertions in DiskUsage checking that values >= 0?

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java [hunk 1]
```diff
@@ -500,7 +500,11 @@ public String toString() {
         }
 
         long getFreeBytes() {
-            return diskUsage.getFreeBytes() - relocatingShardSize;
+            try {
+                return Math.subtractExact(diskUsage.getFreeBytes(), relocatingShardSize);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java]: How could this overflow?
Isn't `relocatingShardSize >= 0` and `0 <= diskUsage.getFreeBytes() <= Long.MAX_VALUE`? Is this to capture the case where `diskUsage.getFreeBytes() == 0 && relocatingShardSize == Long.MAX_VALUE`?

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java]: No, `relocatingShardSize` is the net size of all relocations, i.e. shards in minus shards out, so can indeed be negative :(

**Comment by DaveCTurner** [comment]
I opened #48413 to investigate other sources of negative sizes because it needs a deeper look than I can offer right now.

**Review by ywelsch — APPROVED** [review]
LGTM

**Pull request merged**